segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1435 @@
+            Timer timer = new Timer();
+
+            try (MetadataTransaction txn = metadataStore.beginTransaction()) {
+                // Retrieve metadata and make sure it does not exist.
+                SegmentMetadata oldSegmentMetadata = (SegmentMetadata) txn.get(streamSegmentName);
+                if (null != oldSegmentMetadata) {
+                    throw new StreamSegmentExistsException(streamSegmentName);
+                }
+
+                // Create a new record.
+                SegmentMetadata newSegmentMetatadata = SegmentMetadata.builder()
+                        .name(streamSegmentName)
+                        .maxRollinglength(rollingPolicy.getMaxLength() == 0 ? SegmentRollingPolicy.NO_ROLLING.getMaxLength() : rollingPolicy.getMaxLength())
+                        .ownerEpoch(this.epoch)
+                        .build();
+
+                newSegmentMetatadata.setActive(true);
+                txn.create(newSegmentMetatadata);
+                // commit.
+                txn.commit();
+
+                val retValue = SegmentStorageHandle.writeHandle(streamSegmentName);
+                Duration elapsed = timer.getElapsed();
+                log.debug("{} create - segment={}, rollingPolicy={}, latency={}.", logPrefix, streamSegmentName, rollingPolicy, elapsed.toMillis());
+                LoggerHelpers.traceLeave(log, "create", traceId, retValue);
+                return retValue;
+            } catch (StorageMetadataAlreadyExistsException ex) {
+                throw new StreamSegmentExistsException(streamSegmentName, ex);
+            } catch (StorageMetadataWritesFencedOutException ex) {
+                throw new StorageNotPrimaryException(streamSegmentName, ex);
+            }
+        });
+    }
+
+    /**
+     * Writes the given data to the StreamSegment.
+     *
+     * @param handle  A read-write SegmentHandle that points to a Segment to write to.
+     * @param offset  The offset in the StreamSegment to write data at.
+     * @param data    An InputStream representing the data to write.
+     * @param length  The length of the InputStream.
+     * @param timeout Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate the operation succeeded. If the operation failed,
+     * it will contain the cause of the failure. Notable exceptions:
+     * <ul>
+     * <li> BadOffsetException: When the given offset does not match the actual length of the segment in storage.
+     * <li> StreamSegmentNotExistsException: When the given Segment does not exist in Storage.
+     * <li> StorageNotPrimaryException: When this Storage instance is no longer primary for this Segment (it was fenced out).
+     * </ul>
+     * @throws IllegalArgumentException If handle is read-only.
+     */
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration timeout) {
```

Comment:
I'd not expect a manager abstraction to have critical path operations like reads and writes. Either we are conflating operations that do not belong in the same role or this class should not be called a manager. Why can't we call it simply `ChunkStorage`?

Reply:
ChunkStorageProvider - provides the actual operations on chunks.
ChunkStorageManager - manages the chunks and creates the abstraction of segment out of these chunks. ChunkManager doesn't actually write or read anything. It manages the metadata for segment layout and it maps and forwards calls to ChunkStorageProvider. 

Name ChunkStorage adds ambiguity - does it mean it actually reads/write and then why we have this another thing called ChunkStorageProvider ?

The current name removes that ambiguity.